index.d.ts

```diff
@@ -271,7 +267,7 @@ export interface MiddlewareAPI<S> {
  * asynchronous API call into a series of synchronous actions.
  */
 export interface Middleware {
-  <S>(api: MiddlewareAPI<S>): (next: MiddlewareDispatch) => MiddlewareDispatch;
+  <S>(api: MiddlewareAPI<S>): (next: Dispatch<S>) => any;
```

Comment:
A middleware can return any result. Take a look to `redux-thunk` and `redux-promise` implementations.